File: play_with_moveit/src/play_with_moveit/Ur10eRobot.py
#!/usr/bin/env python

import rospy
import tf
import numpy as np

from moveit_commander import MoveGroupCommander, RobotCommander

from moveit_msgs.msg import MoveItErrorCodes, Grasp, PlaceLocation, RobotTrajectory
from tf.transformations import quaternion_from_euler, quaternion_from_matrix, quaternion_matrix, translation_from_matrix, translation_matrix
from geometry_msgs.msg import WrenchStamped, PoseStamped, PointStamped, Pose
from std_srvs.srv import Empty, EmptyRequest
from std_msgs.msg import Bool
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint

class Ur10eRobot():

    def __init__(self):
        
        self.GRIPPER_GROUP = "ur10e_electric_gripper"
        self.CAMERA_GROUP = "ur10e_camera"
        self.CHANGER_GROUP = "ur10e_tool_changer"
        self.GRIPPER_EF_GROUP = "ur10e_gripper_ef"
        
        self.GRIPPER_EF_LINK = "ur10e_gripper_ef"
        self.CAMERA_EF_LINK = "ur10e_camera_ef"
        self.CHANGER_EF_LINK = "ur10e_changer_ef"
        
        self.___PLANNING_JOINT_NAMES = ["ur10e_shoulder_pan_joint", "ur10e_shoulder_lift_joint", "ur10e_elbow_joint",
                                        "ur10e_wrist_1_joint", "ur10e_wrist_2_joint", "ur10e_wrist_3_joint"]

        self.__GRIPPER_EF_JOINT_NAMES = ["ur10e_upper_finger_base", "ur10e_lower_finger_base"]
        
        self.__FORCE_SENSOR_TOPIC = "/ur10e_force_topic"
        
        # create move group commander
        self.__gripper_group_commander = MoveGroupCommander(self.GRIPPER_GROUP)
        self.__camera_group_commander = MoveGroupCommander(self.CAMERA_GROUP)
        self.__changer_group_commander = MoveGroupCommander(self.CHANGER_GROUP)
        self.__gripper_ef_commander = MoveGroupCommander(self.GRIPPER_EF_GROUP)
        

    """
    Update the current value from force sensor
    """
    def update_force_sensor_value(self, force_sensor_value):
        self.__force_sensor_value = force_sensor_value
        
    
    """
    Get the current value from force sensor
    """
    def get_force_sensor_value(self):
        return self.__force_sensor_value
    
    """
    Open electric gripper
    """
    def open_electric_gripper(self):
        self.__gripper_ef_commander.set_joint_value_target([0, 0])
        self.__gripper_ef_commander.go()
        
    def close_electric_gripper(self, object_width):
        max_width = 0.02
        if (object_width > max_width):
            print ("cannot close gripper, object width exceeds the max width")
        else:
            gripper_dist = (max_width - object_width) / 2
            self.__gripper_ef_commander.set_joint_value_target([-gripper_dist, -gripper_dist])
            self.__gripper_ef_commander.go()
     
    def move_tool_in_straight_line(self, pose_goal, avoid_collisions=True, group_name=None, n_way_points=2):
        
        if (group_name == self.CAMERA_GROUP):
            group = self.__camera_group_commander
        
        elif (group_name == self.CHANGER_GROUP):
            group = self.__changer_group_commander
            
        else:
            group = self.__gripper_group_commander
            
        way_points_list = []
        way_point = Pose()
        x_way_points = np.linspace(0, pose_goal[0], n_way_points)
        y_way_points = np.linspace(0, pose_goal[1], n_way_points)
        z_way_points = np.linspace(0, pose_goal[2], n_way_points)
        qx_way_points = np.linspace(0, 0, n_way_points)
        qy_way_points = np.linspace(0, 0, n_way_points)
        qz_way_points = np.linspace(0, 0, n_way_points)
        qw_way_points = np.linspace(0, 0, n_way_points)
        for i in range(n_way_points):
            way_point.position.x = x_way_points[i]
            way_point.position.y = y_way_points[i]
            way_point.position.z = z_way_points[i]
            way_point.orientation.x = qx_way_points[i]
            way_point.orientation.y = qy_way_points[i]
            way_point.orientation.z = qz_way_points[i]
            way_point.orientation.w = qw_way_points[i]
            way_points_list.append(way_point)
                    
        group.set_pose_reference_frame(group.get_end_effector_link())
        # group.set_goal_tolerance(0.001)

        plan, fraction = group.compute_cartesian_path(way_points_list, 0.005, 0.0)
        print (fraction)
        # print plan

        # post processing for the trajectory to ensure its validity 
        last_time_step = plan.joint_trajectory.points[0].time_from_start.to_sec
        new_plan = RobotTrajectory()
        new_plan.joint_trajectory.header = plan.joint_trajectory.header
        new_plan.joint_trajectory.joint_names = plan.joint_trajectory.joint_names
        new_plan.joint_trajectory.points.append(plan.joint_trajectory.points[0])

        for i in range(1, len(plan.joint_trajectory.points)):
            point = plan.joint_trajectory.points[i]
            if (point.time_from_start.to_sec > last_time_step):
                new_plan.joint_trajectory.points.append(point)
            last_time_step = point.time_from_start.to_sec

        # execute the trajectory
        group.execute(new_plan)
                
    
    """
    Go to pose goal. Plan and execute and wait for the controller respons
    :param pose_goal list in this form [x, y, z, r, p, y]
     or [x, y, z, qx, qy, qz, qw]
     or [x, y, z]
    :type list of int
    :param ef_link
    :type: string
    """
    def go_to_pose_goal(self, pose_goal, group_name=None):
        
        if (group_name == self.CAMERA_GROUP):
            group = self.__camera_group_commander

        elif (group_name == self.CHANGER_GROUP):
            group = self.__changer_group_commander
            
        else:
            group = self.__gripper_group_commander

        if (len(pose_goal) == 6 or len(pose_goal) == 7):
            group.set_pose_target(pose_goal)  
            group.go()                                           

        elif (len(pose_goal) == 3):
            group.set_position_target(pose_goal)  
            group.go() 
            
        else:
            print("Invalid inputs")
        
        
    """
    Go to joint goal
    """
    def go_to_joint_goal(self, joint_goal):
        
        if(len(joint_goal) == 6):
            joint_goal_msg = JointState()
            joint_goal_msg.name = self.___PLANNING_JOINT_NAMES
            joint_goal_msg.position = joint_goal
            self.__camera_group_commander.set_joint_value_target(joint_goal_msg)
            self.__camera_group_commander.go()
        else:
            print("Invalid inputs")
                 
    """
    pick an object using electric gripper
    This function assumes that the electric gripper is already in a place near to the object. With an appropriate orientation.
    So what is left in order to pick. is to approach the object. turn electric gripper on and then retreat.
    """
    def pick_object(self, object_name, table_name, approach_dist, retreat_dist, object_width):
        # disable collision between object and table
        self.__gripper_group_commander.set_support_surface_name(table_name)

        # approach the object
        self.move_tool_in_straight_line(approach_dist, avoid_collisions=False)
        
        # attach object and open gripper
        self.__gripper_group_commander.attach_object(object_name)
        
        # close electric gripper
        self.close_electric_gripper(object_width)
        
        rospy.sleep(2)
        
        # retreat 
        self.move_tool_in_straight_line(retreat_dist, avoid_collisions=True)

    
    """
    Place an object using electric gripper
    """
    def place_object(self, object_name, table_name, approach_dist, retreat_dist):
        # disable collision between object and table
        self.__gripper_group_commander.set_support_surface_name(table_name)

        # approach the table
        self.move_tool_in_straight_line(approach_dist, avoid_collisions=False)
        
        # detach object and turn off gripper
        self.__gripper_group_commander.detach_object(object_name)
        
        # open electric gripper
        self.open_electric_gripper()
        
        rospy.sleep(2)
        
        # retreat 
        self.move_tool_in_straight_line(retreat_dist, avoid_collisions=True)
  
    
    

File: play_with_moveit/src/Ur10eRobot.cpp
#include"play_with_moveit/Ur10eRobot.h"

const std::string Ur10eRobot::GRIPPER_GROUP = "ur10e_electric_gripper";
const std::string Ur10eRobot::GRIPPER_EF_GROUP = "ur10e_gripper_ef";
const std::string Ur10eRobot::CHANGER_GROUP = "ur10e_camera";
const std::string Ur10eRobot::CAMERA_GROUP = "ur10e_tool_changer";
const std::string Ur10eRobot::GRIPPER_EF_LINK = "ur10e_gripper_ef";
const std::string Ur10eRobot::CAMERA_EF_LINK = "ur10e_camera_ef";
const std::string Ur10eRobot::CHANGER_EF_LINK = "ur10e_changer_ef";
const std::vector<std::string> Ur10eRobot::PLANNING_JOINT_NAMES{"ur10e_shoulder_pan_joint", "ur10e_shoulder_lift_joint", "ur10e_elbow_joint",
                                                                "ur10e_wrist_1_joint", "ur10e_wrist_2_joint", "ur10e_wrist_3_joint"};
const std::string Ur10eRobot::FORCE_SENSOR_TOPIC = "/ur10e_force_topic";

Ur10eRobot::Ur10eRobot(ros::NodeHandle *nh)
{
}

Ur10eRobot::~Ur10eRobot()
{
}

geometry_msgs::WrenchStamped Ur10eRobot::getForceSensorValue()
{
    geometry_msgs::WrenchStamped::ConstPtr current_force_value;
    current_force_value = ros::topic::waitForMessage<geometry_msgs::WrenchStamped>(Ur10eRobot::FORCE_SENSOR_TOPIC);
    return *current_force_value;
}
void Ur10eRobot::moveToolInStraightLine(double end_point[3],bool avoid_collisions, std::string group_name)
{

    moveit::planning_interface::MoveGroupInterface group(group_name);
    geometry_msgs::Pose way_point;
    std::vector<geometry_msgs::Pose> way_points_list;

    // push back the starting point always zeros with respect to the tool.
    way_point.position.x = 0;
    way_point.position.y = 0;
    way_point.position.z = 0;
    way_point.orientation.x = 0;
    way_point.orientation.y = 0;
    way_point.orientation.z = 0;
    way_point.orientation.w = 0;
    way_points_list.push_back(way_point);

    // push back the end point
    way_point.position.x = end_point[0];
    way_point.position.y = end_point[1];
    way_point.position.z = end_point[2];
    way_points_list.push_back(way_point);

    // use the required group to plan the cartesian path
    moveit_msgs::RobotTrajectory out_traj;
    double cartesian_success;

    // set the reference frame to the tool
    group.setPoseReferenceFrame(Ur10eRobot::GRIPPER_EF_LINK);

    // plan the cartesian path
    cartesian_success = group.computeCartesianPath(way_points_list, 0.01, 0.0, out_traj, false);

    // process the trajectory to remove invalid points
    out_traj = trajectoryProcessing(out_traj);
    
    // get the current robot state
    moveit_msgs::RobotState robot_state_msg;
    moveit::core::robotStateToRobotStateMsg(*group.getCurrentState(), robot_state_msg);

    // execute the planned path
    const moveit::planning_interface::MoveGroupInterface::Plan cartesian_plan{robot_state_msg, out_traj, 10};
    group.execute(cartesian_plan);
}

void Ur10eRobot::moveRobotTrajectory(double end_point[3],double ur10e_trajectory_points_x[300],double ur10e_trajectory_points_y[300],double ur10e_trajectory_points_z[300],bool avoid_collisions, std::string group_name)
{
    std::cout<<"INSIDE !!!"<<std::endl;
    //std::cout<<end_point[1]<<"\t"<< end_point[2]<<"\t"<<end_point[3]<<std::endl;
    moveit::planning_interface::MoveGroupInterface group(group_name);
    geometry_msgs::Pose way_point;
    std::vector<geometry_msgs::Pose> way_points_list;

    // push back the starting point always zeros with respect to the tool.
    way_point.position.x = 0;
    way_point.position.y = 0;
    way_point.position.z = 0;
    way_point.orientation.x = 0;
    way_point.orientation.y = 0;
    way_point.orientation.z = 0;
    way_point.orientation.w = 0;
    way_points_list.push_back(way_point);

    // push back the end point
    for(int i=1; i<=300;i++)
    {
     way_point.position.x = ur10e_trajectory_points_x[i];
     way_point.position.y = ur10e_trajectory_points_y[i];
     way_point.position.z = ur10e_trajectory_points_z[i];
     way_points_list.push_back(way_point);
    }

    // use the required group to plan the cartesian path
    moveit_msgs::RobotTrajectory out_traj;
    double cartesian_success;

    // set the reference frame to the tool
    group.setPoseReferenceFrame(Ur10eRobot::GRIPPER_EF_LINK);

    // plan the cartesian path
    cartesian_success = group.computeCartesianPath(way_points_list, 0.01, 0.0, out_traj, false);

    // process the trajectory to remove invalid points
    out_traj = trajectoryProcessing(out_traj);
    
    // get the current robot state
    moveit_msgs::RobotState robot_state_msg;
    moveit::core::robotStateToRobotStateMsg(*group.getCurrentState(), robot_state_msg);

    // execute the planned path
    const moveit::planning_interface::MoveGroupInterface::Plan cartesian_plan{robot_state_msg, out_traj, 10};
    group.execute(cartesian_plan);
}

void Ur10eRobot::goToJointGoal(double joint_goal[7])
{
    moveit::planning_interface::MoveGroupInterface group(Ur10eRobot::GRIPPER_GROUP);
    sensor_msgs::JointState joint_msg;
    joint_msg.name = Ur10eRobot::PLANNING_JOINT_NAMES;
    joint_msg.position.assign(joint_goal, joint_goal+7);
    group.setJointValueTarget(joint_msg);
    group.move();
}

void Ur10eRobot::goToPoseGoal(std::vector<double> pose_goal, std::string group_name)
{
    moveit::planning_interface::MoveGroupInterface group(group_name);
    geometry_msgs::Pose pose_msg;

    pose_msg.position.x = pose_goal[0];
    pose_msg.position.y = pose_goal[1];
    pose_msg.position.z = pose_goal[2];

    if (pose_goal.size() == 7)
    {
        pose_msg.orientation.x = pose_goal[3];
        pose_msg.orientation.y = pose_goal[4];
        pose_msg.orientation.z = pose_goal[5];
        pose_msg.orientation.w = pose_goal[6];
    }
    else
    {
        ROS_INFO("invalid input for pose goal");
    }

    group.setPoseTarget(pose_msg);
    group.move();
}
moveit_msgs::RobotTrajectory Ur10eRobot::trajectoryProcessing(moveit_msgs::RobotTrajectory in_traj)
{
    for(int i; i < in_traj.joint_trajectory.points.size()-1; i++)
    {
        if(in_traj.joint_trajectory.points[i+1].time_from_start.toNSec() <= in_traj.joint_trajectory.points[i].time_from_start.toNSec())
        {
            // delete i+1 point
            in_traj.joint_trajectory.points.erase(in_traj.joint_trajectory.points.begin()+i+1);
        }
    }    
    return in_traj;
}
void Ur10eRobot::pickObject(std::string object_name, std::string table_name, double approach_dist[3], double retreat_dist[3])
{
    // initialize  gripper group
    moveit::planning_interface::MoveGroupInterface gripper_group(Ur10eRobot::GRIPPER_GROUP);

    // disable collision betweeen the table and the ef.
    gripper_group.setSupportSurfaceName(table_name);

    // approach the object
    moveToolInStraightLine(approach_dist);

    // attach the object
    gripper_group.attachObject(object_name);

    // close the gripper
    ros::Duration(2).sleep();

    // retreat
    moveToolInStraightLine(retreat_dist);
}
void Ur10eRobot::placeObject(std::string object_name, std::string table_name, double approach_dist[3], double retreat_dist[3])
{
    // initialize gripper group
    moveit::planning_interface::MoveGroupInterface gripper_group(Ur10eRobot::GRIPPER_GROUP);

    // disable collision betweeen the table and the ef.
    gripper_group.setSupportSurfaceName(table_name);

    // approach the place location
    moveToolInStraightLine(approach_dist);

    // detach the object
    gripper_group.detachObject(object_name);

    // open the gripper
    ros::Duration(2).sleep();

    // retreat
    moveToolInStraightLine(retreat_dist);
}
void Ur10eRobot::openGripper()
{
    moveit::planning_interface::MoveGroupInterface group(Ur10eRobot::GRIPPER_EF_GROUP);
    std::vector<double> joint_goal{0.0, 0.0};
    group.setJointValueTarget(joint_goal);
    group.move();
}
void Ur10eRobot::closeGripper(double object_width)
{
    double MAX_WIDTH = 0.02;
    if (object_width > MAX_WIDTH)
        ROS_INFO("cannot close gripper, object width exceeds the max width");
    else
    {
        double gripper_dist = (MAX_WIDTH - object_width) / 2.0;
        moveit::planning_interface::MoveGroupInterface group(Ur10eRobot::GRIPPER_EF_GROUP);
        std::vector<double> joint_goal{-gripper_dist, -gripper_dist};
        group.setJointValueTarget(joint_goal);
        group.move();
    }
}
